src/GcodeLayerThreader.h

```diff
@@ -0,0 +1,213 @@
+/** Copyright (C) 2017 Ultimaker - Released under terms of the AGPLv3 License */
+#ifndef GCODE_LAYER_THREADER_H
+#define GCODE_LAYER_THREADER_H
+
+#include <queue> // priority_queue
+#include <functional> // function
+#include <thread> // sleep
+#include <chrono> // milliseconds
+
+#include "utils/logoutput.h"
+#include "utils/optional.h"
+#include "utils/Lock.h"
+
+namespace cura
+{
+
+/*!
+ * Producer Consumer construct for when:
+ * - production can occur in parallel
+ * - consumption must be ordered and not multithreaded
+ * 
+ * A layer_nr index is passed to the item producer in order to produce the different items.
+ * 
+ * Each thread does production and consumption, giving priority to consumption if some is available.
+ * 
+ * If there is only one thread, it consumes every time it has produced one item.
+ * 
+ * \warning This class is only adequate when the expected production time of an item is more than (n_threads - 1) times as much as the expected consumption time of an item
+ */
+template <typename T>
+class GcodeLayerThreader
+{
+public:
+    /*!
+     * \param start_item_argument_index The first value with which to produce an item
+     * \param end_item_argument_index The last value with which to produce an item
+     * \param produce_item The function with which to produce an item
+     * \param consume_item The function with which to consume an item
+     * \param max_task_count The maximum number of items (being) produced without having been consumed
+     */
+    GcodeLayerThreader(
+        int start_item_argument_index,
+        int end_item_argument_index,
+        const std::function<T* (int)>& produce_item,
+        const std::function<void (T*)>& consume_item,
+        const unsigned int max_task_count
+    )
+    : start_item_argument_index(start_item_argument_index)
+    , end_item_argument_index(end_item_argument_index)
+    , item_count(end_item_argument_index - start_item_argument_index)
+    , max_task_count(max_task_count)
+    , produce_item(produce_item)
+    , consume_item(consume_item)
+    , last_produced_argument_index(start_item_argument_index - 1)
+    {
+        produced.resize(item_count, nullptr);
+    }
```

Comment:
This is a template class. The implementation is going to be in the header file any way, but I could move it outside the class definition to beneath it inside the same header file... Would that be better?